homeassistant/helpers/intent.py

```diff
@@ -117,6 +124,72 @@ def __repr__(self):
         return '<{} - {}>'.format(self.__class__.__name__, self.intent_type)
 
 
+def fuzzymatch(name, entities):
+    """Fuzzy matching function."""
+    matches = []
+    pattern = '.*?'.join(name)
+    regex = re.compile(pattern)
+    for entity in entities:
+        match = regex.search(entity)
+        if match:
+            matches.append((len(match.group()), match.start(), entity))
+    return [x for _, _, x in sorted(matches)]
+
+
+class ServiceIntentHandler(IntentHandler):
+    """Service Intent handler registration.
+
+    Service specific intent handler that calls a service by name/entity_id.
+    """
+
+    slot_schema = {
+        'name': cv.string,
+    }
+
+    def __init__(self, intent_type, domain, service, speech):
+        """Create Service Intent Handler."""
+        self.intent_type = intent_type
+        self.domain = domain
+        self.service = service
+        self.speech = speech
+
+    @asyncio.coroutine
+    def async_handle(self, intent_obj):
+        """Handle the hass intent."""
+        hass = intent_obj.hass
+        slots = self.async_validate_slots(intent_obj.slots)
+        response = intent_obj.create_response()
+
+        name = slots['name']['value']
+        entities = {state.entity_id: state.name for state
+                    in hass.states.async_all()}
+        entity_name = name.replace(' ', '_').lower()
```

Comment:
why ?

Reply:
Comment didn't stick, but sine we convert spaces to _ in friendly names this is needed to match and entity names will never contain a space. But this is better done in the matching code.

Reply:
We match against entity name and not id right?

Reply:
Hmm, should be both since that code should create a dict of id: name but apparently does not for whatever reason and is just creating a list of entity_ids. Flipped it to match on name.